src/BondEvaluatorQuartic.h

```diff
@@ -0,0 +1,241 @@
+    Scalar b_1;         //!< quartic tuning parameter #1
+    Scalar b_2;         //!< quartic tuning parameter #2
+    Scalar U_0;         //!< quartic energy barrier to "breaking"
+    Scalar sigma_6;     //!< sigma raised to the power of 6 for WCA
+    Scalar epsilon_x_4; //!< epsilon * 4 for WCA
+    Scalar delta;       //!< delta parameter bond
+    }
+#if HOOMD_LONGREAL_SIZE == 32
+    __attribute__((aligned(16)));
+#else
+    __attribute__((aligned(32)));
+#endif
+
+//! Class for evaluating the quartic bond potential
+/*!
+ * This bond potential follows the functional form
+ *
+ * \begin{eqnarray*}
+ *      U(r) = & k4 ((r - r_{0}) - b_{1})((r - r_{0}) - b_{2})(r - r_{0})^2
+ *                 + U_{0} + U_{\rm WCA}(r)      & r < r_0\\
+ *            = & U_0                            & r \ge r_0
+ * \end{eqnarray*}
+ *
+ * which has a WCA repulsive force at close range, and a quartic energy well. On the far side of
+ * this energy well, there is an energy barrier, to a plateau energy, U_0. It reaches this energy
+ * at r_0. Beyond r_0, the energy is constant, and there is no force from the bond potential.
+ *
+ * Since this potential has a finite range, beyond which no force is applied, this potential is
+ * useful for modeling bonds with scissile functionality. r_0 defines the bond length at which
+ * the bond can be considered broken, and U_0 defines the height of the energy barrier to breaking.
+ *
+ * The parameters b_1 and b_2 tune the location of the energy minimum, as well as the well width.
+ * They do not cleanly translate to meaningful locations on the potential energy surface, but serve
+ * as helpful tuning parameters to mimic other bond topologies. For example, making a previously
+ * FENE or harmonic bond breakable. Using linear regression or some other fitting algorithm may do
+ * this.
+ *
+ * The delta parameter shifts the whole potential with respect to r. The default value is 0.
+ * Positive values shift to higher r values.
+ */
+class BondEvaluatorQuartic : public BondEvaluator
+    {
+    public:
+    typedef BondParametersQuartic param_type;
+
+    DEVICE BondEvaluatorQuartic(Scalar _rsq, const param_type& _params) : BondEvaluator(_rsq)
+        {
+        k = _params.k;
+        r_0 = _params.r_0;
+        b_1 = _params.b_1;
+        b_2 = _params.b_2;
+        U_0 = _params.U_0;
+        delta = _params.delta;
+        lj1 = _params.epsilon_x_4 * _params.sigma_6 * _params.sigma_6;
+        lj2 = _params.epsilon_x_4 * _params.sigma_6;
+        epsilon = _params.epsilon_x_4 / Scalar(4.0);
+        }
+
+    DEVICE bool evalForceAndEnergy(Scalar& force_divr, Scalar& bond_eng)
+        {
+        bond_eng = 0;
+        force_divr = 0;
+        // check for invalid parameters
+        if (r_0 == Scalar(0.0))
+            return false;
+
+        Scalar r = Scalar(-1.0); // sentinel value for whether the square root has already been
+                                 // taken
```

Comment:
There is alot of duplicated code in this method. Can you restructure it so that isn't the case? Like:
```
force_divr = 0;
bond_eng = 0;

Scalar r_red(1.0);
if (delta == Scalar(0.0))
    {
    r2inv = Scalar(1.0) / rsq;

    // WCA without delta
    bond_eng += ...;
    force_divr += ...;

    if (rsq < r_0 * r_0)
        r_red = fast::sqrt(rsq) - r_0;
    }
else
    {
    r = fast::sqrt(rsq) - delta;

    // WCA with delta
    bond_eng += ...;
    force_divr += ...;

    if (r < r_0)
        r_red = r - r_0;
    }

// bond potential is on when r_red < 0
if (r_red < Scalar(0))
    {
    bond_eng += ...;
    }
else
    {
    bond_eng += U_0;
    }
```
I think the bond potential equations are the same, if you insert delta = 0 in the more general one.